File: src/widget_to_image_converter.c
#include "widget_to_image_converter.h"
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

// Convert RGBA image data to JPEG and save to file
FFI_PLUGIN_EXPORT char *convert_rgba_to_jpeg(const uint8_t *rgba_data,
                                             int width,
                                             int height,
                                             int quality,
                                             const char *output_path)
{
  if (!rgba_data || width <= 0 || height <= 0 || quality < 1 || quality > 100 || !output_path)
  {
    return NULL;
  }

  // Generate unique filename if output_path is a directory
  char final_path[1024];
  if (strlen(output_path) == 0 || output_path[strlen(output_path) - 1] == '/' ||
      output_path[strlen(output_path) - 1] == '\\')
  {
    // Generate timestamp-based filename
    time_t now = time(NULL);
    struct tm *t = localtime(&now);
    snprintf(final_path, sizeof(final_path), "%simage_%04d%02d%02d_%02d%02d%02d.jpg",
             output_path, t->tm_year + 1900, t->tm_mon + 1, t->tm_mday,
             t->tm_hour, t->tm_min, t->tm_sec);
  }
  else
  {
    strncpy(final_path, output_path, sizeof(final_path) - 1);
    final_path[sizeof(final_path) - 1] = '\0';
  }

  // Convert RGBA to RGB (stb_image_write expects RGB)
  int rgb_size = width * height * 3;
  uint8_t *rgb_data = malloc(rgb_size);
  if (!rgb_data)
    return NULL;

  for (int i = 0; i < width * height; i++)
  {
    rgb_data[i * 3] = rgba_data[i * 4];         // R
    rgb_data[i * 3 + 1] = rgba_data[i * 4 + 1]; // G
    rgb_data[i * 3 + 2] = rgba_data[i * 4 + 2]; // B
  }

  // Write JPEG file
  int success = stbi_write_jpg(final_path, width, height, 3, rgb_data, quality);
  free(rgb_data);

  if (!success)
    return NULL;

  // Return allocated copy of the path
  char *result = malloc(strlen(final_path) + 1);
  if (result)
  {
    strcpy(result, final_path);
  }

  return result;
}

// Convert RGBA file to JPEG and save to file
FFI_PLUGIN_EXPORT char *convert_rgba_file_to_jpeg(const char *input_path,
                                                  int width,
                                                  int height,
                                                  int quality,
                                                  const char *output_path)
{
  if (!input_path || width <= 0 || height <= 0 || quality < 1 || quality > 100 || !output_path)
  {
    return NULL;
  }

  // Open and read the input RGBA file
  FILE *input_file = fopen(input_path, "rb");
  if (!input_file)
  {
    return NULL;
  }

  // Get file size
  fseek(input_file, 0, SEEK_END);
  long file_size = ftell(input_file);
  fseek(input_file, 0, SEEK_SET);

  if (file_size <= 0)
  {
    fclose(input_file);
    return NULL;
  }

  // Calculate expected file size based on dimensions
  long expected_size = width * height * 4;
  if (file_size != expected_size)
  {
    fclose(input_file);
    return NULL;
  }

  // Read RGBA data
  uint8_t *rgba_data = malloc(file_size);
  if (!rgba_data)
  {
    fclose(input_file);
    return NULL;
  }

  size_t bytes_read = fread(rgba_data, 1, file_size, input_file);
  fclose(input_file);

  if (bytes_read != file_size)
  {
    free(rgba_data);
    return NULL;
  }

  // Generate unique filename if output_path is a directory
  char final_path[1024];
  if (strlen(output_path) == 0 || output_path[strlen(output_path) - 1] == '/' ||
      output_path[strlen(output_path) - 1] == '\\')
  {
    // Generate timestamp-based filename
    time_t now = time(NULL);
    struct tm *t = localtime(&now);
    snprintf(final_path, sizeof(final_path), "%simage_%04d%02d%02d_%02d%02d%02d.jpg",
             output_path, t->tm_year + 1900, t->tm_mon + 1, t->tm_mday,
             t->tm_hour, t->tm_min, t->tm_sec);
  }
  else
  {
    strncpy(final_path, output_path, sizeof(final_path) - 1);
    final_path[sizeof(final_path) - 1] = '\0';
  }

  // Convert RGBA to RGB (stb_image_write expects RGB)
  int rgb_size = width * height * 3;
  uint8_t *rgb_data = malloc(rgb_size);
  if (!rgb_data)
  {
    free(rgba_data);
    return NULL;
  }

  for (int i = 0; i < width * height; i++)
  {
    rgb_data[i * 3] = rgba_data[i * 4];         // R
    rgb_data[i * 3 + 1] = rgba_data[i * 4 + 1]; // G
    rgb_data[i * 3 + 2] = rgba_data[i * 4 + 2]; // B
  }

  // Write JPEG file
  int success = stbi_write_jpg(final_path, width, height, 3, rgb_data, quality);

  // Free allocated memory
  free(rgba_data);
  free(rgb_data);

  if (!success)
    return NULL;

  // Return allocated copy of the path
  char *result = malloc(strlen(final_path) + 1);
  if (result)
  {
    strcpy(result, final_path);
  }

  return result;
}

File: lib/src/widget_to_image_converter.dart
// You have generated a new plugin project without specifying the `--platforms`
// flag. An FFI plugin project that supports no platforms is generated.
// To add platforms, run `flutter create -t plugin_ffi --platforms <platforms> .`
// in this directory. You can also find a detailed instruction on how to
// add platforms in the `pubspec.yaml` at
// https://flutter.dev/to/pubspec-plugin-platforms.

import 'dart:ffi';
import 'dart:io';
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import 'widget_to_image_converter_bindings_generated.dart';

const String _libName = 'widget_to_image_converter';

/// The dynamic library in which the symbols for [WidgetToImageConverterBindings] can be found.
final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

/// The bindings to the native functions in [_dylib].
final WidgetToImageConverterBindings _bindings = WidgetToImageConverterBindings(
  _dylib,
);

/// Convert RGBA image data to JPEG and save to file
///
/// Parameters:
///   rgbaData: RGBA image data (4 bytes per pixel: R, G, B, A)
///   width: Image width in pixels
///   height: Image height in pixels
///   quality: JPEG quality (1-100)
///   outputPath: Path where to save the JPEG file
///
String convertRgbaToJpeg(
  Uint8List rgbaData,
  int width,
  int height,
  int quality,
  String outputPath,
) {
  // Validate parameters
  if (rgbaData.isEmpty) {
    throw ArgumentError('RGBA data cannot be empty');
  }
  if (width <= 0 || height <= 0) {
    throw ArgumentError('Width and height must be positive');
  }
  if (quality < 1 || quality > 100) {
    throw ArgumentError('Quality must be between 1 and 100');
  }
  if (outputPath.isEmpty) {
    throw ArgumentError('Output path cannot be empty');
  }

  // Check if data size matches dimensions
  final expectedSize = width * height * 4;
  if (rgbaData.length != expectedSize) {
    throw ArgumentError(
      'RGBA data size (${rgbaData.length}) does not match dimensions '
      '(${width}x${height}x4 = $expectedSize)',
    );
  }

  // Allocate memory for RGBA data
  final Pointer<Uint8> rgbaPtr = calloc<Uint8>(rgbaData.length);
  rgbaPtr.asTypedList(rgbaData.length).setAll(0, rgbaData);

  // Convert output path to C string
  final Pointer<Char> outputPathPtr = outputPath.toNativeUtf8().cast<Char>();

  // Call C function
  final Pointer<Char> resultPtr = _bindings.convert_rgba_to_jpeg(
    rgbaPtr,
    width,
    height,
    quality,
    outputPathPtr,
  );

  // Free allocated memory
  calloc.free(rgbaPtr);
  calloc.free(outputPathPtr);

  // Check if conversion was successful
  if (resultPtr == nullptr) {
    throw Exception('Failed to convert RGBA to JPEG');
  }

  // Get the actual path where the file was saved
  final String actualPath = resultPtr.cast<Utf8>().toDartString();

  // Free the result pointer (allocated in C)
  calloc.free(resultPtr);

  return actualPath;
}

/// Convert RGBA file to JPEG and save to file
///
/// Parameters:
///   inputPath: Path to the input RGBA file
///   width: Image width in pixels
///   height: Image height in pixels
///   quality: JPEG quality (1-100)
///   outputPath: Path where to save the JPEG file
///
String convertRgbaFileToJpeg(
  String inputPath,
  int width,
  int height,
  int quality,
  String outputPath,
) {
  // Validate parameters
  if (inputPath.isEmpty) {
    throw ArgumentError('Input path cannot be empty');
  }
  if (width <= 0 || height <= 0) {
    throw ArgumentError('Width and height must be positive');
  }
  if (quality < 1 || quality > 100) {
    throw ArgumentError('Quality must be between 1 and 100');
  }
  if (outputPath.isEmpty) {
    throw ArgumentError('Output path cannot be empty');
  }

  // Convert paths to C strings
  final Pointer<Char> inputPathPtr = inputPath.toNativeUtf8().cast<Char>();
  final Pointer<Char> outputPathPtr = outputPath.toNativeUtf8().cast<Char>();

  // Call C function
  final Pointer<Char> resultPtr = _bindings.convert_rgba_file_to_jpeg(
    inputPathPtr,
    width,
    height,
    quality,
    outputPathPtr,
  );

  // Free allocated memory
  calloc.free(inputPathPtr);
  calloc.free(outputPathPtr);

  // Check if conversion was successful
  if (resultPtr == nullptr) {
    throw Exception('Failed to convert RGBA file to JPEG');
  }

  // Get the actual path where the file was saved
  final String actualPath = resultPtr.cast<Utf8>().toDartString();

  // Free the result pointer (allocated in C)
  calloc.free(resultPtr);

  return actualPath;
}
